lib/scrub_output_structure.rb

```diff
@@ -0,0 +1,159 @@
+# frozen_string_literal: true
+
+require "pathname"
+require "json"
+
+# Sets up a directory structure for storing autoscrub output files:
+#
+# data/
+# - member_data/
+# - - x/
+# - - - log/
+# - - - output/
+#
+# ... where x is the member_id.
+# Subdirs (loaded, log, output, ready_to_load) are generated upon init.
+# They can have dated subdirs.
+#
+# - - - loaded/
+# - - - log/
+# - - - - 2020-01-01/
+# - - - - 2020-02-01/
+# - - - output/
+# - - - - 2020-01-01/
+# - - - - 2020-02-01/
+# - - - ready_to_load/
+#
+# Create dated subdirs for member x (using current date):
+#
+# sos = ScrubOutputStructure.new("xyz")
+#
+# sos.date_subdir!("output") # -> Dir
+# sos.date_subdir!("log")    # -> Dir
+#
+# Get latest exising subdir for log or output:
+#
+# sos.latest("log")    # -> Dir
+# sos.latest("output") # -> Dir
+class ScrubOutputStructure
+  VALID_SUBDIR   = ["log", "output", "ready_to_load", "loaded"].freeze
+  VALID_DATE_STR = /^\d\d\d\d-\d\d-\d\d$/.freeze
+  DATE_STR       = /\d\d\d\d-\d\d-\d\d/.freeze
+
+  attr_reader :member_id, :member_dir, :member_log, :member_output,
+              :member_ready_to_load, :member_loaded
+
+  def initialize(member_id)
+    unless member_id.is_a?(String)
+      raise "member_id must be a string"
+    end
+
+    @member_id     = member_id
+    @member_dir    = mkbase!
+    @member_log    = mklog!
+    @member_output = mkoutput!
+    @member_ready_to_load = mkready_to_load!
+    @member_loaded = mkloaded!
+  end
+
+  # A call like:
+  # ScrubOutputStructure.new("xyz").date_subdir!("log")
+  # ... on 2020-12-30 makes the dir
+  # /usr/src/app/data/member_data/xyz/log/2020-12-30/
+  def date_subdir!(subdir, date_str = Time.new.strftime("%F"))
+    mkdir!(
+      validate_subdir(subdir),
+      validate_date_str(date_str)
+    )
+  end
+
+  # Remove the dir that belongs to @member_id and everything under it.
+  def remove_recursive!
+    FileUtils.remove_dir(member_dir.path)
+    true
+  end
+
+  def latest(subdir)
+    base_dir      = mkdir!(validate_subdir(subdir))
+    latest_subdir = base_dir.children.select {|str| str.match?(DATE_STR) }.max
+
+    if latest_subdir.nil?
+      nil
+    else
+      Dir.new(File.join(base_dir, latest_subdir))
+    end
+  end
+
+  def to_json(*_args)
+    dir_hash = {
+      "member_id"            => member_id,
```

Comment:
If we find ourselves wanting to add more kinds of subdirectories we might want to refactor to specify the kinds of subdirectories only once rather than in multiple places in this class; no need to change that now.

Reply:
Yeah I can imagine we'll at least be adding something for shared print here.